libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
         return this;
     }
 
+    /**
+     * Used in {@link #writableFieldAsBase64(String, WritableValue)}
+     */
+    @FunctionalInterface
+    public interface WritableValue {
+        void writeTo(OutputStream os) throws IOException;
+    }
+
+    /**
+     * Write the serialization of a {@link WritableValue} via {@link WritableValue#writeTo(OutputStream)} as a string encoded in
+     * Base64 format. This API can be used to generate XContent directly without the intermediate results to reduce memory usage.
+     */
+    public XContentBuilder writableFieldAsBase64(String name, WritableValue value) throws IOException {
+        generator.writeFieldDirectly(name, os -> {
+            os.write('\"');
+            try (OutputStream encoded = Base64.getEncoder().wrap(new FilterOutputStream(os) {
+                @Override
+                public void close() {
+                    // do not close the output
+                }
+            })) {
+                value.writeTo(encoded);
+            }
```

Comment:
Do these quotes presume you are writing to json? I imagine there is an content type neutral way to do this but I don't know it off hand.